GNOME 3 needs boltd for Thunderbolt 3 support

## Issue description

As described in [this blog post](https://christian.kellner.me/2017/12/14/introducing-bolt-thunderbolt-3-security-levels-for-gnulinux/), GNOME 3 needs the [bolt daemon](https://gitlab.freedesktop.org/bolt/bolt) to handle authenticating Thunderbolt 3 devices.

This is not yet packaged for NixOS.

I've [attempted to package](https://gist.github.com/callahad/c9808a9bb85d8dcc0338d93eeef0bae0) bolt myself but have thus far been unsuccessful.

### Steps to reproduce

- Start a Gnome 3 session
- Navigate to Settings → Devices → Thunderbolt

__Expected Results:__ A nice GUI as [described here](https://christian.kellner.me/2018/04/23/the-state-of-thunderbolt-3-in-fedora-28/).

__Actual Results:__ An error message:

![screenshot of the error message transcribed below](https://user-images.githubusercontent.com/24193/52911932-4a0d7f00-32a2-11e9-99ae-f9fe9409dc6d.png)

> No Thunderbolt support
>
> Thunderbolt could not be detected.
> Either the system lacks Thunderbolt support, it has been disabled in the BIOS, or is set to an unsupported security level in the BIOS.

(Note: The error message is slightly misleading. Thunderbolt 3 *is* supported and working on my laptop, albeit with authorization disabled, but without `bolt` GNOME doesn't know that.)

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.20.7, NixOS, 19.03pre169108.36f31600749 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.03pre169108.36f31600749"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Right now I'm hitting two roadblocks with bolt itself:

1. Building the man pages fails somewhere inside asciidoc:

```
[21/94] Generating boltctl-man with a custom command.
FAILED: boltctl.1
/nix/store/nh2pywspgv2m7sn9c8nhjb59rs31x0pm-asciidoc-8.6.9/bin/a2x -d manpage -f manpage -D . -a 'manmanual=bolt Manual' -a mansource=bolt -a revnumber=0.7 ../docs/boltctl.1.txt
a2x: WARNING: --destination-dir option is only applicable to HTML based outputs
a2x: ERROR: "xmllint" --nonet --noout --valid "/build/source/build/boltctl.1.xml" returned non-zero exit status 4
```

I'm not sure how to get more information / further work on that.

2. If I disable manpage generation (`mesonFlags = [ "-Dman=false" ]`), everything compiles and links just fine but then fails with:

```
[91/92] Linking target test-udev.
[92/92] Linking target test-sysfs.
glibPreInstallPhase
glibPreInstallPhase
installing
install flags: install
[0/1] Installing files.
pkexec must be setuid root
FAILED: meson-install
/nix/store/y6jq0jwxhzjackxp6bsyh93v9s6bnpa6-meson-0.49.0/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
builder for '/nix/store/clxfzrq236dw0sqc6imvn0lqmpbyzcfb-bolt-0.7.drv' failed with exit code 1
error: build of '/nix/store/clxfzrq236dw0sqc6imvn0lqmpbyzcfb-bolt-0.7.drv' failed
```

--- a/maintainers/maintainer-list.nix
+++ b/maintainers/maintainer-list.nix
@@ -752,6 +752,11 @@
     github = "calbrecht";
     name = "Christian Albrecht";
   };
+  callahad = {
+    email = "dan.callahan@gmail.com";
+    github = "callahad";
+    name = "Dan Callahan";
+  };
   calvertvl = {
     email = "calvertvl@gmail.com";
     github = "calvertvl";

--- a/nixos/modules/module-list.nix
+++ b/nixos/modules/module-list.nix
@@ -280,6 +280,7 @@
   ./services/hardware/acpid.nix
   ./services/hardware/actkbd.nix
   ./services/hardware/bluetooth.nix
+  ./services/hardware/bolt.nix
   ./services/hardware/brltty.nix
   ./services/hardware/freefall.nix
   ./services/hardware/fwupd.nix

--- a/nixos/modules/services/hardware/bolt.nix
+++ b/nixos/modules/services/hardware/bolt.nix
@@ -1,0 +1,34 @@
+# Thunderbolt 3 device manager
+
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+{
+  options = {
+
+    services.hardware.bolt = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Bolt, a userspace daemon to enable
+          security levels for Thunderbolt 3 on GNU/Linux.
+
+          Bolt is used by GNOME 3 to handle Thunderbolt settings.
+        '';
+      };
+
+    };
+
+  };
+
+  config = mkIf config.services.hardware.bolt.enable {
+
+    environment.systemPackages = [ pkgs.bolt ];
+    services.udev.packages = [ pkgs.bolt ];
+    systemd.packages = [ pkgs.bolt ];
+
+  };
+}

--- a/nixos/modules/services/x11/desktop-managers/gnome3.nix
+++ b/nixos/modules/services/x11/desktop-managers/gnome3.nix
@@ -151,6 +151,7 @@ in {
     services.colord.enable = mkDefault true;
     services.packagekit.enable = mkDefault true;
     hardware.bluetooth.enable = mkDefault true;
+    services.hardware.bolt.enable = mkDefault true;
     services.xserver.libinput.enable = mkDefault true; # for controlling touchpad settings via gnome control center
     services.udev.packages = [ pkgs.gnome3.gnome-settings-daemon ];
     systemd.packages = [ pkgs.gnome3.vino ];

--- a/pkgs/development/python-modules/python-dbusmock/default.nix
+++ b/pkgs/development/python-modules/python-dbusmock/default.nix
@@ -1,0 +1,59 @@
+{ lib, buildPythonPackage, fetchPypi,
+  nose, dbus, dbus-python, pygobject3,
+  which, pyflakes, pycodestyle, bluez, networkmanager
+}:
+
+buildPythonPackage rec {
+  pname = "python-dbusmock";
+  version = "0.18.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1hj02p65cic4jdc6a5xf1hx8j5icwy7dcrm5kg91lkjks4gwpg5h";
+  };
+
+  prePatch = ''
+    sed -i -e 's|pyflakes3|pyflakes|g' tests/test_code.py;
+  '';
+
+  # TODO: Get the rest of these tests running?
+  # This is a mocking library used as a check dependency for a single derivation.
+  # That derivation's tests pass. Maybe not worth the effort to fix these...
+  NOSE_EXCLUDE = lib.concatStringsSep "," [
+    "test_bluez4" # NixOS ships BlueZ5
+    # These appear to fail because they're expecting to run in an Ubuntu chroot?
+    "test_everything" # BlueZ5 OBEX
+    "test_polkitd"
+    "test_consolekit"
+    "test_api"
+    "test_logind"
+    "test_notification_daemon"
+    "test_ofono"
+    "test_gnome_screensaver"
+    "test_cli"
+    "test_timedated"
+    "test_upower"
+    # Very slow, consider disabling?
+    # "test_networkmanager"
+  ];
+
+  checkInputs = [
+    nose dbus dbus-python which pycodestyle pyflakes
+    pygobject3 bluez bluez.test networkmanager
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+    export PATH="$PATH:${bluez.test}/test";
+    nosetests -v
+    runHook postCheck
+  '';
+
+  meta = with lib; {
+    description = "Mock D-Bus objects for tests";
+    homepage = https://github.com/martinpitt/python-dbusmock;
+    license = licenses.lgpl3Plus;
+    maintainers = with maintainers; [ callahad ];
+    platforms = platforms.linux;
+  };
+}

--- a/pkgs/os-specific/linux/bolt/0001-skip-mkdir.patch
+++ b/pkgs/os-specific/linux/bolt/0001-skip-mkdir.patch
@@ -1,0 +1,12 @@
+diff --git a/scripts/meson-install.sh b/scripts/meson-install.sh
+index 859ae81..05a1c58 100644
+--- a/scripts/meson-install.sh
++++ b/scripts/meson-install.sh
+@@ -7,5 +7,5 @@ fi
+ 
+ BOLT_DBDIR=$1
+ 
+-echo "Creating database dir: ${BOLT_DBDIR}"
+-mkdir -p "${DESTDIR}/${BOLT_DBDIR}"
++# echo "Creating database dir: ${BOLT_DBDIR}"
++# mkdir -p "${DESTDIR}/${BOLT_DBDIR}"

--- a/pkgs/os-specific/linux/bolt/default.nix
+++ b/pkgs/os-specific/linux/bolt/default.nix
@@ -1,0 +1,61 @@
+{ stdenv, meson, ninja, pkgconfig, fetchFromGitLab,
+  python3, umockdev, gobject-introspection, dbus,
+  asciidoc, libxml2, libxslt, docbook_xml_dtd_45, docbook_xsl,
+  glib, systemd, polkit
+}:
+
+stdenv.mkDerivation rec {
+  pname = "bolt";
+  version = "0.7";
+
+  src = fetchFromGitLab {
+    domain = "gitlab.freedesktop.org";
+    owner = "bolt";
+    repo = "bolt";
+    rev = "${version}";
+    sha256 = "0xn2c31kcjh1j76gq1qrcxwjyjyqnsxygkfrvh3xk07qc92f99xd";
+  };
+
+  nativeBuildInputs = [
+    meson ninja pkgconfig
+    asciidoc libxml2 libxslt docbook_xml_dtd_45 docbook_xsl
+  ] ++ stdenv.lib.optional (!doCheck) python3;
+
+  buildInputs = [
+    glib systemd polkit
+  ];
+
+  doCheck = true;
+
+  preCheck = ''
+    export LD_LIBRARY_PATH=${umockdev.out}/lib/
+  '';
+
+  checkInputs = [
+    dbus umockdev gobject-introspection
+    (python3.withPackages
+      (p: [ p.pygobject3 p.dbus-python p.python-dbusmock ]))
+  ];
+
+  # meson install tries to create /var/lib/boltd
+  patches = [ ./0001-skip-mkdir.patch ];
+
+  postPatch = ''
+    patchShebangs tests/test-integration
+  '';
+
+  mesonFlags = [
+    "-Dlocalstatedir=/var"
+  ];
+
+  PKG_CONFIG_SYSTEMD_SYSTEMDSYSTEMUNITDIR = "${placeholder "out"}/lib/systemd/system";
+  PKG_CONFIG_UDEV_UDEVDIR = "${placeholder "out"}/lib/udev";
+
+  meta = with stdenv.lib; {
+    description = "Thunderbolt 3 device management daemon";
+    homepage = https://gitlab.freedesktop.org/bolt/bolt;
+    license = licenses.lgpl21Plus;
+    maintainers = [ maintainers.callahad ];
+    platforms = platforms.linux;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -14314,6 +14314,8 @@ in
 
   inherit (python3Packages) bedup;
 
+  bolt = callPackage ../os-specific/linux/bolt { };
+
   bridge-utils = callPackage ../os-specific/linux/bridge-utils { };
 
   busybox = callPackage ../os-specific/linux/busybox { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -720,6 +720,8 @@ in {
 
   python-binance = callPackage ../development/python-modules/python-binance { };
 
+  python-dbusmock = callPackage ../development/python-modules/python-dbusmock { };
+
   python-engineio = callPackage ../development/python-modules/python-engineio { };
 
   python-hosts = callPackage ../development/python-modules/python-hosts { };

